epsilon_greedy.go

```diff
@@ -143,45 +140,55 @@ func (p *epsilonGreedyHostPool) getEpsilonGreedy() string {
 		return p.getRoundRobin()
 	}
 
-	// calculate values for each host in the 0..1 range (but not ormalized)
-	var possibleHosts []*hostEntry
+	// here we are exploiting
+	var (
+		totalScore            float64
+		candidateEpsilonHosts = make([]*hostEntry, 0, 256)
+		candidateScores       = make([]float64, 0, 256)
+	)
+
 	now := time.Now()
-	var sumValues float64
 	for _, h := range p.hostList {
-		if h.canTryHost(now) {
-			v := h.getWeightedAverageResponseTime()
-			if v > 0 {
-				ev := p.CalcValueFromAvgResponseTime(v)
-				h.epsilonValue = ev
-				sumValues += ev
-				possibleHosts = append(possibleHosts, h)
-			}
+		if !h.canTryHost(now) {
+			continue
 		}
-	}
 
-	if len(possibleHosts) != 0 {
-		// now normalize to the 0..1 range to get a percentage
-		for _, h := range possibleHosts {
-			h.epsilonPercentage = h.epsilonValue / sumValues
```

Comment:
I got rid of this `epsilonPercentage` value. It's a local value to this calculation so has no reason being part of that struct. It's unclean and also means in future it's as difficult to get rid of the mutual lock